[Bug] ProtobufNativeSchemaTest.testSchema order-independent

### Search before reporting

- [x] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Read release policy

- [x] I understand that [unsupported versions](https://pulsar.apache.org/contribute/release-policy/#supported-versions) don't get bug fixes. I will attempt to reproduce the issue on a supported version of Pulsar client and Pulsar broker.


### User environment

Broker version: 4.2.0-SNAPSHOT
Broker Operating system and hardware type: Linux x86_64 GNU/Linux
Broker Java version: 17.0.16


### Issue Description

In ProtobufNativeSchemaTest.testSchema, the json's contents and the FileDescriptorSet field's contents do not have a deterministic order but the hardcoded string assertion assumes a deterministic order. The json serializer did not guarantee attribute order and inside FileDescriptorSet the contents can also be in different orders due to different generation paths or environments producing the contents in different orders despite the logical content being the same. Since the test compares the raw strings/trees "as-is", harmless re-ordering could flip the test from pass to fail without any real schema change.

I discovered the problem with the [NonDex](https://github.com/TestingResearchIllinois/NonDex) tool. NonDex systematically detects incorrect tests that rely on non-deterministic behaviors in Java APIs—like assuming order of name/value pairs in json files — by exploring all specification-allowed outcomes. It does this by instrumenting undetermined APIs and randomizing the returned order/behavior within what the spec allows. Failures it exposes reliably indicate flawed assumptions in order that were never promised and should be fixed.

### Error messages

```text
[ERROR] Failures: 
[ERROR] org.apache.pulsar.client.impl.schema.ProtobufNativeSchemaTest.testSchema
[INFO]   Run 1: PASS
[ERROR]   Run 2: ProtobufNativeSchemaTest.testSchema:71 expected [{"fileDescriptorSet":"CtMDCgpUZXN0LnByb3RvEgVwcm90bxoSRXh0ZXJuYWxUZXN0LnByb3RvImUKClN1Yk1lc3NhZ2USCwoDZm9vGAEgASgJEgsKA2JhchgCIAEoARo9Cg1OZXN0ZWRNZXNzYWdlEgsKA3VybBgBIAEoCRINCgV0aXRsZRgCIAEoCRIQCghzbmlwcGV0cxgDIAMoCSLlAQoLVGVzdE1lc3NhZ2USEwoLc3RyaW5nRmllbGQYASABKAkSEwoLZG91YmxlRmllbGQYAiABKAESEAoIaW50RmllbGQYBiABKAUSIQoIdGVzdEVudW0YBCABKA4yDy5wcm90by5UZXN0RW51bRImCgtuZXN0ZWRGaWVsZBgFIAEoCzIRLnByb3RvLlN1Yk1lc3NhZ2USFQoNcmVwZWF0ZWRGaWVsZBgKIAMoCRI4Cg9leHRlcm5hbE1lc3NhZ2UYCyABKAsyHy5wcm90by5leHRlcm5hbC5FeHRlcm5hbE1lc3NhZ2UqJAoIVGVzdEVudW0SCgoGU0hBUkVEEAASDAoIRkFJTE9WRVIQAUItCiVvcmcuYXBhY2hlLnB1bHNhci5jbGllbnQuc2NoZW1hLnByb3RvQgRUZXN0YgZwcm90bzMKoAEKEkV4dGVybmFsVGVzdC5wcm90bxIOcHJvdG8uZXh0ZXJuYWwiOwoPRXh0ZXJuYWxNZXNzYWdlEhMKC3N0cmluZ0ZpZWxkGAEgASgJEhMKC2RvdWJsZUZpZWxkGAIgASgBQjUKJW9yZy5hcGFjaGUucHVsc2FyLmNsaWVudC5zY2hlbWEucHJvdG9CDEV4dGVybmFsVGVzdGIGcHJvdG8z","rootMessageTypeName":"proto.TestMessage","rootFileDescriptorName":"Test.proto"}] but found [{"rootFileDescriptorName":"Test.proto","rootMessageTypeName":"proto.TestMessage","fileDescriptorSet":"CtMDCgpUZXN0LnByb3RvEgVwcm90bxoSRXh0ZXJuYWxUZXN0LnByb3RvImUKClN1Yk1lc3NhZ2USCwoDZm9vGAEgASgJEgsKA2JhchgCIAEoARo9Cg1OZXN0ZWRNZXNzYWdlEgsKA3VybBgBIAEoCRINCgV0aXRsZRgCIAEoCRIQCghzbmlwcGV0cxgDIAMoCSLlAQoLVGVzdE1lc3NhZ2USEwoLc3RyaW5nRmllbGQYASABKAkSEwoLZG91YmxlRmllbGQYAiABKAESEAoIaW50RmllbGQYBiABKAUSIQoIdGVzdEVudW0YBCABKA4yDy5wcm90by5UZXN0RW51bRImCgtuZXN0ZWRGaWVsZBgFIAEoCzIRLnByb3RvLlN1Yk1lc3NhZ2USFQoNcmVwZWF0ZWRGaWVsZBgKIAMoCRI4Cg9leHRlcm5hbE1lc3NhZ2UYCyABKAsyHy5wcm90by5leHRlcm5hbC5FeHRlcm5hbE1lc3NhZ2UqJAoIVGVzdEVudW0SCgoGU0hBUkVEEAASDAoIRkFJTE9WRVIQAUItCiVvcmcuYXBhY2hlLnB1bHNhci5jbGllbnQuc2NoZW1hLnByb3RvQgRUZXN0YgZwcm90bzMKoAEKEkV4dGVybmFsVGVzdC5wcm90bxIOcHJvdG8uZXh0ZXJuYWwiOwoPRXh0ZXJuYWxNZXNzYWdlEhMKC3N0cmluZ0ZpZWxkGAEgASgJEhMKC2RvdWJsZUZpZWxkGAIgASgBQjUKJW9yZy5hcGFjaGUucHVsc2FyLmNsaWVudC5zY2hlbWEucHJvdG9CDEV4dGVybmFsVGVzdGIGcHJvdG8z"}]
[INFO] 
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
INFO: Surefire failed when running tests for c8A8RmGSuO8XYxjfW87EPw3D18O8xdB4vMgX7ULah0=
[INFO] NonDex SUMMARY:
[INFO] *********
[INFO] mvn nondex:nondex  -DnondexFilter='.*' -DnondexMode=FULL -DnondexSeed=974622 -DnondexStart=0 -DnondexEnd=9223372036854775807 -DnondexPrintstack=false -DnondexDir="/home/leby2/pulsar/pulsar-client/.nondex" -DnondexJarDir="/home/leby2/pulsar/pulsar-client/.nondex" -DnondexExecid=c8A8RmGSuO8XYxjfW87EPw3D18O8xdB4vMgX7ULah0= -DnondexLogging=CONFIG
[WARNING] org.apache.pulsar.client.impl.schema.ProtobufNativeSchemaTest#testSchema
[INFO] *********
[INFO] All tests pass without NonDex shuffling
[INFO] ####################
[INFO] Across all seeds:
[INFO] org.apache.pulsar.client.impl.schema.ProtobufNativeSchemaTest#testSchema
[INFO] Test results can be found at: 
[INFO] file:///home/leby2/pulsar/pulsar-client/.nondex/c8A8RmGSuO8XYxjfW87EPw3D18O8xdB4vMgX7ULah0=/test_results.html
[INFO] [NonDex] The id of this run is: c8A8RmGSuO8XYxjfW87EPw3D18O8xdB4vMgX7ULah0=
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  26.577 s
[INFO] Finished at: 2025-10-01T10:30:04-05:00
[INFO] ------------------------------------------------------------------------
[INFO] 0 goals, 0 executed
[ERROR] Failed to execute goal edu.illinois:nondex-maven-plugin:2.2.1:nondex (default-cli) on project pulsar-client-original: Unable to execute mojo: There are test failures.
```

### Reproducing the issue

You can replicate the problem with the [NonDex](https://github.com/TestingResearchIllinois/NonDex) tool command:
```mvn -pl pulsar-client -Dtest=org.apache.pulsar.client.impl.schema.ProtobufNativeSchemaTest#testSchema -DforkCount=1 -DreuseForks=false -DnondexSeed=974622 -DnondexRuns=1 edu.illinois:nondex-maven-plugin:2.2.1:nondex```

### Additional information

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaTest.java
+++ b/pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/ProtobufNativeSchemaTest.java
@@ -20,14 +20,25 @@ package org.apache.pulsar.client.impl.schema;
 
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertNotNull;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.protobuf.DescriptorProtos;
 import com.google.protobuf.Descriptors;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufAllocator;
 import java.nio.charset.StandardCharsets;
+import java.util.Base64;
 import java.util.Collections;
 import java.util.HashMap;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.pulsar.common.schema.SchemaType;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.skyscreamer.jsonassert.JSONAssert;
+import org.skyscreamer.jsonassert.JSONCompareMode;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
@@ -46,6 +57,28 @@ public class ProtobufNativeSchemaTest {
             + "gBQjUKJW9yZy5hcGFjaGUucHVsc2FyLmNsaWVudC5zY2hlbWEucHJvdG9CDEV4dGVybmFsVGVzdGIGcHJvdG8z\",\"rootMessageT"
             + "ypeName\":\"proto.TestMessage\",\"rootFileDescriptorName\":\"Test.proto\"}";
 
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    /**
+     * Decode a base64-encoded Protobuf FileDescriptorSet into a canonical JSON tree.
+     * @param b64 a base64 string
+     * @return a normalized JSON tree that can be compared with JSONAssert, ensuring deterministic
+     * equality checks regardless of field ordering.
+     * @throws IllegalArgumentException if b64 isn't valid Base64
+     * @throws InvalidProtocolBufferException if the decoded bytes are not a valid FileDescriptorSet
+     * @throws JSONException if the JSON string is invalid
+     */
+    private static JSONObject fdsToJson(String b64)
+            throws IllegalArgumentException, InvalidProtocolBufferException, JSONException {
+        DescriptorProtos.FileDescriptorSet fds =
+            DescriptorProtos.FileDescriptorSet.parseFrom(Base64.getDecoder().decode(b64));
+        String json = JsonFormat.printer()
+            .includingDefaultValueFields()
+            .omittingInsignificantWhitespace()
+            .print(fds);
+        return new JSONObject(json);
+    }
+
     @Test
     public void testEncodeAndDecode() {
         final String stringFieldValue = "StringFieldValue";
@@ -61,15 +94,49 @@ public class ProtobufNativeSchemaTest {
     }
 
     @Test
-    public void testSchema() {
+    public void testSchema() throws Exception {
         ProtobufNativeSchema<org.apache.pulsar.client.schema.proto.Test.TestMessage> protobufSchema =
                 ProtobufNativeSchema.of(org.apache.pulsar.client.schema.proto.Test.TestMessage.class);
 
         assertEquals(protobufSchema.getSchemaInfo().getType(), SchemaType.PROTOBUF_NATIVE);
 
         assertNotNull(ProtobufNativeSchemaUtils.deserialize(protobufSchema.getSchemaInfo().getSchema()));
-        assertEquals(new String(protobufSchema.getSchemaInfo().getSchema(),
-                StandardCharsets.UTF_8), EXPECTED_SCHEMA_JSON);
+
+        // Parse the actual/expected JSON into trees
+        String actualJson   = new String(protobufSchema.getSchemaInfo().getSchema(), StandardCharsets.UTF_8);
+        JsonNode actualRoot   = MAPPER.readTree(actualJson);
+        JsonNode expectedRoot = MAPPER.readTree(EXPECTED_SCHEMA_JSON);
+
+        // Extract and validate the FileDescriptorSet field for semantic comparison
+        // (When decoded, Protobuf descriptors can serialize fields in varying orders
+        // causing hard coded string comparisons to fail)
+        String fdSetField = "fileDescriptorSet";
+        JsonNode actualB64Node = actualRoot.path(fdSetField);
+        JsonNode expectedB64Node = expectedRoot.path(fdSetField);
+        Assert.assertFalse(actualB64Node.isMissingNode());
+        Assert.assertFalse(expectedB64Node.isMissingNode());
+        Assert.assertTrue(actualB64Node.isValueNode());
+        Assert.assertTrue(expectedB64Node.isValueNode());
+
+        // Decode FileDescriptorSets to JSON and compare semantically (order-insensitive)
+        JSONObject actualFdsObj   = fdsToJson(actualB64Node.asText());
+        JSONObject expectedFdsObj = fdsToJson(expectedB64Node.asText());
+        JSONAssert.assertEquals(
+                "FileDescriptorSet mismatch: decoded Protobuf descriptors have mismatched schema contents",
+                expectedFdsObj,
+                actualFdsObj,
+                JSONCompareMode.NON_EXTENSIBLE
+        );
+
+        // Remove the already verified field and compare remaining schema attributes, order does not matter
+        ((ObjectNode) actualRoot).remove(fdSetField);
+        ((ObjectNode) expectedRoot).remove(fdSetField);
+        JSONAssert.assertEquals(
+                "Schema metadata mismatch: remaining JSON fields differ after verifying FileDescriptorSet",
+                MAPPER.writeValueAsString(expectedRoot),
+                MAPPER.writeValueAsString(actualRoot),
+                JSONCompareMode.NON_EXTENSIBLE
+        );
     }
 
     @Test

